subsys: bluetooth: host: hci_core: Fix bt_le_set_chan_map

adamfc2000 opened the pull request:
Allow the channel map to be set when periodic advertising or acting as an ISO source.

adamfc2000 committed:
Bluetooth: Host: Fix bt_le_set_chan_map

Allow the channel map to be set when periodic advertising or acting as
an ISO source.

Signed-off-by: Adam Cavender <adam.cavender@nordicsemi.no>

hermabe commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4276,7 +4276,10 @@ int bt_le_set_chan_map(uint8_t chan_map[5])
+	if (!(IS_ENABLED(CONFIG_BT_CENTRAL)
+	    | IS_ENABLED(CONFIG_BT_ISO_BROADCASTER)
+	    | IS_ENABLED(CONFIG_BT_ISO_CENTRAL)
+	    | IS_ENABLED(CONFIG_BT_PER_ADV))) {
```
With the "Channel Classification Enhancement" added in 5.3 it may be useful for the ACL peripheral to set the channel map as well, we should maybe add that one while we are at it?

jhedberg replied:
Are these really supposed to be bitwise or and not logical or?

Thalley replied:
It should definitely be `||` instead of `|` :)

adamfc2000 replied:
> Are these really supposed to be bitwise or and not logical or?

Oops, my mistake!

hermabe reviewed (commented):
The ifdefs for `cmd_chan_map` in the shell should also be updated (removed?) so that the commands work for the new roles: https://github.com/zephyrproject-rtos/zephyr/blob/1b364c1422010dd450146510237d11779395b85c/subsys/bluetooth/shell/bt.c#L3060 https://github.com/zephyrproject-rtos/zephyr/blob/1b364c1422010dd450146510237d11779395b85c/subsys/bluetooth/shell/bt.c#L4107-L4110

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4276,7 +4276,10 @@ int bt_le_set_chan_map(uint8_t chan_map[5])
 	struct net_buf *buf;
 
-	if (!IS_ENABLED(CONFIG_BT_CENTRAL)) {
+	if (!(IS_ENABLED(CONFIG_BT_CENTRAL)
+	    | IS_ENABLED(CONFIG_BT_ISO_BROADCASTER)
+	    | IS_ENABLED(CONFIG_BT_ISO_CENTRAL)
+	    | IS_ENABLED(CONFIG_BT_PER_ADV))) {
```
I wonder if this could be simplified as `IS_ENABLED(CONFIG_BT_CENTRAL) || IS_ENABLED(CONFIG_BT_BROADCASTER)` as all the higher layer roles (such as `CONFIG_BT_ISO_BROADCASTER` and `CONFIG_BT_ISO_CENTRAL`) will depend on either of those.

adamfc2000 committed:
Bluetooth: Shell: Fix cmd_chan_map

Allow the channel map to be set when periodic advertising or acting as
an ISO source.

Signed-off-by: Adam Cavender <adam.cavender@nordicsemi.no>

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -4104,7 +4104,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(bt_cmds,
 		      cmd_conn_phy_update, 2, 3),
 #endif
-#if defined(CONFIG_BT_CENTRAL)
+#if defined(CONFIG_BT_CENTRAL) || defined(CONFIG_BT_BROADCASTER)
 	SHELL_CMD_ARG(channel-map, NULL, "<channel-map: XXXXXXXXXX> (36-0)",
 		      cmd_chan_map, 2, 1),
 #endif /* CONFIG_BT_CENTRAL */
```
Since the function itself is not guarded (but will just return an error), do we need the guard here at all?

hermabe replied:
Is it not guarded by `CONFIG_BT_CENTRAL`? https://github.com/zephyrproject-rtos/zephyr/blob/3d29c9fe546cd42554becee9dddcbe2e9e73d5cc/subsys/bluetooth/shell/bt.c#L3236-L3237

Damian-Nordic replied:
@adamfc2000
`cmd_chan_map` is not built if only CONFIG_BT_BROADCASTER is enabled (and CONFIG_BT_CENTRAL is not). Could this be fixed?
